### PR TITLE
Project

### PR DESCRIPTION
**Pull request by supreme5678**
Project list and Project detail page

### CODE REVIEW

File: web/src/components/TaskOverlay.tsx [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+      alert('You can only upload JPG/PNG file!')
+    }
+    const isLt2M = file.size / 1024 / 1024 < 2
+    if (!isLt2M) {
+      alert('Image must smaller than 2MB!')
+    }
+    return isJpgOrPng && isLt2M
+  }
+
+  const handleChange = (info: any) => {
```
Comment (tonklaartisan) [inline web/src/components/TaskOverlay.tsx:67]: function ?

File: web/src/components/TaskOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    onCloseModal()
+  }
+  function getBase64(img: any, callback: any) {
+    const reader = new FileReader()
+    reader.addEventListener('load', () => callback(reader.result))
+    reader.readAsDataURL(img)
+  }
+
+  function beforeUpload(file: any) {
+    const isJpgOrPng = file.type === 'image/jpeg' || file.type === 'image/png'
```
Comment (tonklaartisan) [inline web/src/components/TaskOverlay.tsx:56]: isValid ?

File: web/src/components/TaskOverlay.tsx [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+} from '@ant-design/icons'
+import { Button, Col, Input, Modal, Row, Skeleton, Tooltip, Typography, Upload } from 'antd'
+import Avatar from 'antd/lib/avatar/avatar'
+import React, { useEffect, useState } from 'react'
+import { Link } from 'react-router-dom'
+
+function TaskOverlay({ data, project, visible, onCloseModal }: any) {
+  const { Text } = Typography
+
+  const [taskData, setTaskData]: any = useState({})
```
Comment (tonklaartisan) [inline web/src/components/TaskOverlay.tsx:23]: const [taskData, setTaskData]: Task | null = useState(null)

File: web/src/components/TaskOverlay.tsx [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+      userImg: 'https://source.unsplash.com/600x600/?cat',
+      userName: 'John Doe',
+      message,
+    }
+    let tempData = { ...taskData, comments: [...taskData.comments, tempComment] }
+    setTaskData(tempData)
+    setMessage('')
+  }
+
+  return Object.keys(taskData).length === 0 ? (
```
Comment (tonklaartisan) [inline web/src/components/TaskOverlay.tsx:97]: !taskData ? (<no task>) : (<task>)

File: web/src/pages/ProjectList.tsx [hunk 5]
```diff
@@ -1,18 +1,100 @@
...
+
+  function handleKeywordChange(e: any) {
+    setLoading(true)
+    setKeyword(e.target.value)
+    if (e.target.value === '') {
+      setFilteredData(DATA)
+      setTypes('all')
+      setLoading(false)
+    } else {
+      let kw: any[] = DATA.filter((item) => {
```
Comment (tonklaartisan) [inline web/src/pages/ProjectList.tsx:46]: const

**Review by akoranos — APPROVED** [review]

File: web/src/components/TaskCard.tsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  function openModal() {
+    setModalVisible(true)
+  }
+
+  function onDoneClick(event: any) {
+    event.stopPropagation()
+
+    let tempData = { ...taskData }
```
Comment (tonklaartisan) [inline web/src/components/TaskCard.tsx:32]: So weird, I think we can improve this.

File: web/src/components/TaskOverlay.tsx [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+  const [taskData, setTaskData] = useState<Task | null>(null)
+  const [modalVisible, setModalVisible] = useState(false)
+  const [imageUrl, setImageUrl] = useState()
+  const [loading, setLoading] = useState(false)
+  const [message, setMessage] = useState('')
+
+  console.log('imageUrl', imageUrl)
+  console.log('taskData', taskData)
+  console.log('data', data)
```
Comment (tonklaartisan) [inline web/src/components/TaskOverlay.tsx:31]: Should remove console.log before commit

File: web/src/components/TaskCard.tsx (hunk 6 above)
Comment (supreme5678) [reply web/src/components/TaskCard.tsx:32]: stopPropagation?

File: web/src/pages/ProjectList.tsx [hunk 8]
```diff
@@ -1,18 +1,100 @@
...
+      case 'all':
+        setFilteredData(DATA)
+        break
+      default:
+        setFilteredData(DATA)
+        break
+    }
+  }, [types]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  function handleTypeChange(e: any) {
```
Comment (tonklaartisan) [inline web/src/pages/ProjectList.tsx:34]: If a function has one line, we can remove it by using an inline function `<Radio.Group className="my-4" value={types} onChange={(e) => setTypes(e.target.value)}>` instead.

**Review by tonklaartisan — COMMENTED** [review]
- Should remove `console.log` before commit
- Should use short return in .map() in TSX
- Try to reduce `:any` type, for TSX, see https://react-typescript-cheatsheet.netlify.app/docs/advanced/types_react_api/
- This PR has some conflicts with `develop`, should rebase and resolve first

File: web/src/components/TaskCard.tsx (hunk 6 above)
Comment (tonklaartisan) [reply web/src/components/TaskCard.tsx:32]: let tempData = { ...taskData }

**Pull request merged**